File: Project Οντοκεντρικός Προγραμματισμός/C++ Project/CodeBlocks Project Hotel/src/Room.cpp
#include <Room.h>
#include <Booking.h>

int Room::roomInstanceCounter = 100;//Initialization of the variable roomInstanceCounter at 100, used for Room ID binding.
int pricePerPerson;//Variable used for the price per person of a Room object.
int maxPeople;//Variable used for the maximum people of a Room object.
Booking *roomAvailability[31];//An 31 Booking pointers elements array.
int roomID;//Variable used for Room ID binding.
static int roomInstanceCounter;//Variable used for Room ID binding.

Room::Room()//Constructor for a Room type object.
{
	setPricePerPerson(15);//Initialization of the pricePerPerson variable at 15.
	setMaxPeople(10);//Initialization of the maxPeople variable at 10.
	Booking *roomAvailability[31] = { 0 };//Initialization of every element of the roomAvailability array at 0(null pointer).
	setRoomID();//Increasing the roomID variable by each successful room constructed.
}

Room::Room(int pricePerPerson)//Constructor for a Room type object.
{
	setPricePerPerson(pricePerPerson);//Initialization of the pricePerPerson variable at 15.
	setMaxPeople(10);//Initialization of the maxPeople variable at 10.
	Booking *roomAvailability[31] = { 0 };//Initialization of every element of the roomAvailability array at 0(null pointer).
	setRoomID();//Increasing the roomID variable by each successful room constructed.
}

Room::~Room()//Default destructor for a Room type object.
{
	roomInstanceCounter--;
}

bool Room::addBooking(Booking* a)//Function that returns true if you can book in a room and books at it or false if not.
{
	int i;
	for (i = a->getArrivalDay(); i <= a->getArrivalDay()+ a->getAccomodationDays() - 1; i++)
	{
		if (this->roomAvailability[i] != 0 || a->getPeopleNum() > this->getMaxPeople())
			return false;
	}
	for (i = a->getArrivalDay(); i <= a->getArrivalDay() + a->getAccomodationDays() - 1; i++)
	{
		this->roomAvailability[i] = a;
	}
	a->setRoom(this);//Change the Room type pointer form 0(null pointer) to this Room.
	return true;
}


double Room::getPriceOfRoom()//Function that returns the money earned from a room.
{
	int i;
	double moneyEarnedFromRoom = 0;
	for (i = 1; i < 31; i++)
	{
		if (this->roomAvailability[i] != 0)
			moneyEarnedFromRoom += (roomAvailability[i]->getPeopleNum() * getPricePerPerson());
	}
	return moneyEarnedFromRoom;
}


bool Room::cancelBooking(int BookingID)//Function that cancels a booking.
{
	int i;
	for (i = 1; i<31; i++)
	{
		if (this->roomAvailability[i] != 0 && this->roomAvailability[i]->getBookingID() == BookingID)
		{
			this->roomAvailability[i] = 0;
		}
	}
	return true;
}


double Room::checkFullness()//Function that returns the percentage fullness of a room.
{
	int i;
	double fullnessPercentage = 0;
	int daysCounter = 0;
	for (i = 1; i<31; i++)
	{
		if (this->roomAvailability[i] != 0)
		{
			daysCounter++;
		}
	}
	fullnessPercentage = ((double)daysCounter / 30) * 100.0;
	return fullnessPercentage;
}

int getPricePerPerson()//Getter for the variable pricePerPerson.
{
	return pricePerPerson;
}

void setPricePerPerson(int val)//Setter for the variable pricePerPerson.
{
	pricePerPerson = val;
}

int getMaxPeople()//Getter for the variable maxPeople.
{
	return maxPeople;
}

void setMaxPeople(int val)//Setter for the variable maxPeople.
{
	maxPeople = val;
}

void setRoomID()//Setter for the variable roomID.
{
	roomID = ++roomInstanceCounter;
}

int getRoomID()//Getter for the variable roomID.
{
	return roomID;
}

Booking* getRoomAvailability(int i)//Function that returns the element of the roomAvailability array at [i].
{
	return *(roomAvailability + i);
}

File: Project Οντοκεντρικός Προγραμματισμός/Java Project/Hotel Project, Swing/Room.java
public class Room //Η κλάση για ένα γενικού τύπου δωμάτιο του ξενοδοχείου.
{
    private static int roomInstanceCounter=100; //Μεταβλητή για την αντιστοίχιση του ID.
    private int roomID;                         //Μεταβλητή για την αντιστοίχιση του ID.
    private static int maxPeople=10;            //Μεταβλητή για μέγιστο αριθμό ατόμων.
    private static double pricePerPerson=15;    //Μεταβλητή για τιμή ανά άτομο.

    protected Booking roomAvailability[] = new Booking[31];
     
    //-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    //Setters
    public void setRoomID() 
    {
        roomInstanceCounter++;
        roomID = roomInstanceCounter;
    }
    public static void setMaxPeople(int maxPeoplee)
    {
        maxPeople=maxPeoplee;
    }
    public static void setPricePerPerson(int pricePerPersonn)
    {
        pricePerPerson=pricePerPersonn;
    }
    //Getters
    public double getPricePerPerson()
    {
        return pricePerPerson;
    }
     public int getMaxPeople()
    {
        return maxPeople;
    }
    public int getRoomID()
    {
        return roomID;
    }
    //--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    
    public Room()//Constructor για το αντικείμενο Room.
    {
        int i=0;
        Booking roomAvailability[] = new Booking[31];
        for(i=0; i < 31; i++)
        {
            roomAvailability[i]=null;
        }
        setRoomID();//Κάθε φορά που δημιουργούμε ένα δωμάτιο καλείται η μέθοδος setRoomID() για την αντιστοίχιση δωματίου και ID.
    }
    public boolean addBooking(Booking a) //Η μέθοδος ελέγχει αν μπορεί να γίνει κράτηση σε ένα δωμάτιο,αν ταιριάζουν
    {                                    //οι μέρες διαμονής και η μέρα άφιξης και το προσθέτει στον πίνακα διαθεσιμότητας.
        int i=0; 
        for(i=a.getArrivalDate(); i < a.getArrivalDate()+a.getAccomodationDays()-1; i++)
        {
             if(roomAvailability[i]!=null || a.getPeopleNum()>maxPeople)
            {
                return false;
            }
        }
        for(i=a.getArrivalDate(); i <= a.getArrivalDate()+a.getAccomodationDays()-1; i++)
        {
            roomAvailability[i] = a;
        }
        a.setRoom(this);//Μέθοδος που ενημερώνει την μεταβλητή Room στο πεδίο της Booking.
        return true;
    }
    public double getPriceOfRoom()//Μέθοδος υπολογισμού εσόδων ενός δωματίου.
    {
        int i=0;
        double moneyEarnedFromRoom; //Μεταβλητή για έσοδα δωματίου.
        moneyEarnedFromRoom=0;
        for(i=1; i<31; i++)
        {
            if(roomAvailability[i]!=null)
                {
                    moneyEarnedFromRoom = (moneyEarnedFromRoom + (roomAvailability[i].getPeopleNum() * getPricePerPerson()));
                }
        }
        return moneyEarnedFromRoom;
    }
    public boolean cancelBooking(int bookingID)//Μέθοδος ακύρωσης συγκεκριμένης κράτησης σε ένα δωμάτιο.
    {
        int i;
        boolean found=false;
        for(i=1; i<31; i++)
        {
            if( this.roomAvailability[i]!=null && roomAvailability[i].getBookingID() == bookingID)
            {
                roomAvailability[i]=null;
                found=true;
                
            }
        }
            return true;
    }
    public double checkFullness()//Μέθοδος η οποία εκτυπώνει την διαθεσημότητα ενός δωματίου.
    {
        int i;
        double fullnessPercentage=0;//Μεταβλητή στην οποία αποθηκεύεται το ποσοστό.
        int daysCounter=0;//Μεταβλητή που μετράει τις μέρες και την αξιοποιούμε για να βρούμε το ποσοστό.
        for(i=1; i<31; i++)//Αν κάθε κελί του πίνακα διαθεσιμότητας δεν είναι null τότε αυξάνει τον daysCounter κατά 1.
        {
            if(roomAvailability[i]!=null)
            {
                daysCounter++;
            }
        }
        fullnessPercentage=((double)daysCounter/30) * 100.0;//Υπολογισμός ποσοστού.
        return fullnessPercentage;
    }
}

    